### ISSUE TITLE
Puppetdb facts lookup doesn't work with arrays

### ISSUE THREAD

**Issue by gmenuel** [puppetlabs/bolt#3065]
## Describe the Bug
In the Puppetdb plugin documentation it is written that the module support PQL dot notation for querying facts, but it doesn't seems to work with arrays.

## Expected Behavior

Trying to query a fact like `facts.az_metadata.network.interface[0].ipv4.ipAddress[0].publicIpAddress` should work. I also tried with facts.az_metadata.network.interface.0.ipv4.ipAddress.0.publicIpAddress but the same problem occurs.

## Steps to Reproduce

For reproducing this bug on any installation I use the facts.processors.models[0] entry, this is of course useless in real life usage.
Create a inventory.yaml like :

```
groups:
  - name: test
    targets:
    - _plugin: puppetdb
      query: "inventory[certname] { } "
      target_mapping:
        name: certname
        config:
          ssh:
            user: facts.processors.models[0]
```

This should return stuff like "user: Intel(R) Xeon(R) CPU..."
```
bolt inventory show --detail |grep user:
```

## Environment
 - Bolt 3.21.0
- PuppetDB  7.9.0
- Debian 11

**Comment by donoghuc** [comment]
Thanks for the detailed report! Yeah this is a bug. It looks like one option is to massage how we build up our query to puppetdb. I hacked together a patch like this:
```ruby
+ cd pe-modules-2021.5.0.0.24.ge6b65142
+ exit 0
diff --git a/lib/bolt/plugin/puppetdb.rb b/lib/bolt/plugin/puppetdb.rb
index 7a1dd996..2fd5f720 100644
--- a/lib/bolt/plugin/puppetdb.rb
+++ b/lib/bolt/plugin/puppetdb.rb
@@ -42,10 +42,21 @@ module Bolt

       def fact_path(raw_fact)
         fact_path = raw_fact.split(".")
-        if fact_path[0] == 'facts'
-          fact_path.drop(1)
+        # pull out index for puppetdb query ex my_fact[0] becomes [my_fact, 0]
+        index_expression = /(?<index>\[\d+\]$)/
+        expanded_fact_path = []
+        fact_path.each do |segment|
+          if match = segment.match(index_expression)
+            expanded_fact_path << segment.gsub(index_expression, '')
+            expanded_fact_path << match['index'][1..-2].to_i
+          else
+            expanded_fact_path << segment
+          end
+        end
+        if expanded_fact_path[0] == 'facts'
+          expanded_fact_path.drop(1)
         elsif fact_path == ['certname']
-          fact_path
+          expanded_fact_path
         else
           raise FactLookupError.new(raw_fact, "fact lookups must start with 'facts.'")
         end
@@ -75,11 +86,9 @@ module Bolt
           end
           value
         end
-
         facts.uniq!
         # Returns {'mycertname' => [{'path' => ['nested', 'fact'], 'value' => val'}], ... }
         fact_values = @puppetdb_client.fact_values(targets, facts)
-
         targets.map do |certname|
           target_data = fact_values[certname]
           target = resolve_facts(template, certname, target_data) || {}
(END)
```
That appears to solve the issue, but... it brings in a kind of messy regular expression. I'll mention this to the team and figure out how we want to fix this for real. Cheers.

**Comment by donoghuc** [comment]
Actually, I totally forgot to even look in to this, but the way to index in to an array is to represent it with the index using dot notation (no square brackets) So like:
```yaml
groups:
  - name: test
    targets:
    - _plugin: puppetdb
      query: "inventory[certname] { } "
      target_mapping:
        name: certname
        config:
          ssh:
            user: facts.processors.models.0
```

This makes fixing the bug trivial: (something like)
```diff
diff --git a/lib/bolt/plugin/puppetdb.rb b/lib/bolt/plugin/puppetdb.rb
index 7a1dd996..336a582d 100644
--- a/lib/bolt/plugin/puppetdb.rb
+++ b/lib/bolt/plugin/puppetdb.rb
@@ -42,6 +42,15 @@ module Bolt

       def fact_path(raw_fact)
         fact_path = raw_fact.split(".")
+        fact_path = fact_path.map do |segment|
+          begin
+            # Turn it into an integer if we can
+            Integer(segment)
+          rescue ArgumentError
+            # Otherwise return the value
+            segment
+          end
+        end
         if fact_path[0] == 'facts'
           fact_path.drop(1)
         elsif fact_path == ['certname']
@@ -75,11 +84,9 @@ module Bolt
           end
           value
         end
-
         facts.uniq!
         # Returns {'mycertname' => [{'path' => ['nested', 'fact'], 'value' => val'}], ... }
         fact_values = @puppetdb_client.fact_values(targets, facts)
-
         targets.map do |certname|
           target_data = fact_values[certname]
           target = resolve_facts(template, certname, target_data) || {}
```

**Comment by gmenuel** [comment]
> but the way to index in to an array is to represent it with the index using dot notation (no square brackets)

That's not actually what is written in the doc, [bolt documentation](https://puppet.com/docs/bolt/latest/supported_plugins.html#puppetdb) says to use the PQL dot notation with a link to [PQL doc](https://puppet.com/docs/puppetdb/7/api/query/v4/ast.html#dot-notation):

> Array indexing (inventory only) is represented with brackets ([]) on the end of a key.

I prefer the puppet dot notation (with .0 instead of [0]) but if we want to change this we need to update the doc too.

**Comment by donoghuc** [comment]
I think that while the doc is somewhat confusing, it is accurate. It notes that brackets are for inventory only endpoint in pdb. The puppetdb plugin usese the `query` endpoint https://github.com/puppetlabs/bolt/blob/f76f7444a1fbf44da413062e4a8e9102ccced66a/lib/bolt/puppetdb/client.rb#L67 . I believe that https://puppet.com/docs/bolt/latest/supported_plugins.html#available-fact-paths the note that 

> [PQL dot notation](https://puppet.com/docs/puppetdb/latest/api/query/v4/ast.html#dot-notation) facts string such as facts.os.family to reference fact value. Dot notation is required for both structured and unstructured facts.

was meant to indicate that dots (not brackets) are acceptable, linking to the docs there assumes the reader knows that under the hood the plugin is using the `query` not `inventory` endpoint. 

I will just update the docs as part of my PR and we will see if I can come up with something that is more clear.

**Comment by beechtom** [comment]
Closed by #3066
